File: algorithms/01_python/math/cholesky.py
"""
Cholesky LL^T Decomposition in Python (Numerical Recipes 3rd Ed. Chapter 2.6).
"""

import math

def cholesky(A):
    n = len(A)
    L = [[0.0] * n for _ in range(n)]

    for i in range(n):
        for j in range(i + 1):
            s = sum(L[i][k] * L[j][k] for k in range(j))
            if i == j:
                val = A[i][i] - s
                if val <= 0:
                    raise ValueError("Matrix not positive definite")
                L[i][j] = math.sqrt(val)
            else:
                L[i][j] = (A[i][j] - s) / L[j][j]
    return L

if __name__ == "__main__":
    A = [
        [4, 12, -16],
        [12, 37, -43],
        [-16, -43, 98]
    ]
    L = cholesky(A)
    assert abs(L[0][0] - 2.0) < 1e-6
    print("Python Cholesky verified.")

File: algorithms/19_lua/math/cholesky.lua
-- Cholesky LL^T in Lua (Numerical Recipes 3rd Ed. Chapter 2.6)

local function cholesky(A)
    local n = #A
    local L = {}
    for i = 1, n do
        L[i] = {}
        for j = 1, n do L[i][j] = 0.0 end
    end

    for i = 1, n do
        for j = 1, i do
            local sum = 0.0
            for k = 1, j - 1 do
                sum = sum + L[i][k] * L[j][k]
            end
            if i == j then
                local val = A[i][i] - sum
                assert(val > 0, "Not positive definite")
                L[i][j] = math.sqrt(val)
            else
                L[i][j] = (A[i][j] - sum) / L[j][j]
            end
        end
    end
    return L
end

local A = {
    {4.0, 12.0, -16.0},
    {12.0, 37.0, -43.0},
    {-16.0, -43.0, 98.0}
}
local L = cholesky(A)
assert(math.abs(L[1][1] - 2.0) < 1e-6)
print("Lua Cholesky verified.")
